docs/content/preview/architecture/docdb-replication/raft.md

```diff
@@ -0,0 +1,74 @@
+
+YugabyteDB's implementation of RAFT is based on [Apache Kudu](https://kudu.apache.org/docs/)'s implementation of RAFT but has several enhancements, such as leader leases and pre-voting state during learner mode for correctness, improvements to the network stack, auto balancing of tablets on failures, zone/DC aware data placement, leader-balancing, ability to do full cluster moves in a online manner, and more.
+
+Let us go over the different component involved in the RAFT consenses protocol and how it works.
+
+{{<tip>}}
+To see an animated visual of the working of RAFT, see [RAFT Visualization](https://raft.github.io/)
+{{</tip>}}
+
+## Roles
+
+In RAFT, nodes in the distributed system can take on one of three roles.
+
+- **Leader** : The leader is responsible for managing the replicated log and coordinating the other nodes. The Leader continues to send heartbeat messages to maintain its leadership status.
+- **Follower** : The leader replicates data in the form of log entries to follower nodes. One of the followers becomes the leader when the leader fails.
+- **Candidate** : A follower transitions to the Candidate state, to start a new election term and attempts to become the new leader.
+
+Initially, all nodes start as followers.
+
+## Log entries
+
+Log entries represent the sequence of commands or state changes that need to be replicated and applied across all nodes in the distributed system. Each node maintains its own log, which is an ordered sequence of log entries. Log entries play a fundamental role in achieving consensus and ensuring fault tolerance.
+
+## Log replication
+
+The leader is responsible for replicating log entries to the followers. When a client sends a request to the leader, the leader appends the request to its local log and then sends the log entry to the followers. The followers then append the log entry to their local logs and send an acknowledgment back to the leader. Once the leader receives a majority of acknowledgments, it considers the log entry to be committed and can apply it to its state machine.
+
+## Replication of the write operation
+
+All write operations are handled by the tablet leader. These are the squence of operations that happen during the replication of a write.
+
+1. The leader first takes a lock on the key using a local in-memory lock manager and reads data if necessary (for read-modify-write or conditional update operations). Note that this locking mechanism does not exist on followers.
```

Comment:
actually many locks have to be taken as we have to lock the prefixes usually
for snapshot isolation there's also some validation required to make sure nothing has already been written between the read time and the current time (simplifying)

 this is also assuming single-shard transaction's so no intents have to be written as part of the locking
 we do however, have to check for intents from other distributed transactions regardless